---
title: "The R package **bigstatsr**:<br/>Memory- and Computation-Efficient Tools for Big Matrices"
subtitle: "useR!2017 lightning talk"
author: "Florian Priv√© (@privefl)"
date: "July 6, 2017"
output:
  xaringan::moon_reader:
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: true
      # autoplay: 20000
---

```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(echo = FALSE, fig.align = 'center')
```

## About

I'm a PhD Student (2016-2019) in **Predictive Human Genetics** in Grenoble.

$$\boxed{\Large{\text{Disease} \sim \text{DNA mutations}}}$$

```{r}
knitr::include_graphics("http://www.blacksheep-van.com/wp-content/uploads/2016/03/Blacksheep-location-de-van-ame%CC%81nage-a%CC%80-Grenoble-1.jpg")
```

---

## Very large genotype matrices

- currently: 15K x 300K, [celiac disease](http://www.nature.com/ng/journal/v42/n4/abs/ng.543.html)

- soon: 500K x 800K, [UK Biobank](https://doi.org/10.1371/journal.pmed.1001779)
 
```{r, out.width='65%'}
knitr::include_graphics("https://media.giphy.com/media/3o7bueyxGydy48Lwgo/giphy.gif")
```

---

## Problem I had

```{r, out.width='100%'}
knitr::include_graphics("memory-problem.svg")
```

---

## Solution I found

```{r, out.width='100%'}
knitr::include_graphics("memory-solution.svg")
```

.footnote[Michael J. Kane, John Emerson, Stephen Weston (2013).]

---

## Accesses almost as if the matrix were in memory

- in **R**: accesses with `[`, as standard R matrices,

- in **Rcpp**: accesses single elements with `X(i, j)`, as standard Rcpp matrices.

<br/>

```{r, out.width='60%'}
knitr::include_graphics("https://media.giphy.com/media/Rl9Yqavfj2Ula/giphy.gif")
```

---

## Split-(par)Apply-Combine Strategy

### Apply standard R functions to big matrices (in parallel)

```{r, out.width='100%'}
knitr::include_graphics("split-apply-combine.svg")
```

.footnote[strategy coined by Hadley Wickham (2011)]

---

## Matrix operations

- (cross-)products with matrices/vectors

- special tricks for handling scaling ([vignette](https://privefl.github.io/bigstatsr/articles/operations-with-scaling.html) and [blog post](https://goo.gl/L8cNbo))

<br/>

### Example: computation of correlation of a 100,000 x 5000 matrix

- `cor`: 22 minutes 

- `big_cor`: 1 minute 

---

## Sparse linear models: **biglasso**

```{r, out.width='70%'}
knitr::include_graphics("https://raw.githubusercontent.com/YaohuiZeng/biglasso/master/vignettes/2016-11-20_vary_p_pkgs.png")
```

.footnote[Zeng, Y., and Breheny, P. (2017).]

---

## Partial Singular Value Decomposition

### Guess: 

If X is 15K x **300K** `big.matrix`, how much time takes 

```{r, eval=FALSE, echo=TRUE}
big_randomSVD(X, big_scale(), k = 10, ncores = 6)
```

### Hint: 

If X is 15K x **30K**

```{r, eval = FALSE, echo=TRUE}
svd(scale(X), nu = 10, nv = 10)
```

takes 50 minutes and 18GB of RAM.

---

## Partial SVD took only **1 minute**

```{r, out.width='70%'}
knitr::include_graphics("celiac-pca.png")
```

.footnote[based on R package **RSpectra**]

---

## Test association of each variable with an outcome

### In genetics, this is called a Genome-Wide Association Study (GWAS)

<br/>

```{r, out.width='100%', fig.cap="Manhattan plot"}
knitr::include_graphics("https://upload.wikimedia.org/wikipedia/commons/1/12/Manhattan_Plot.png")
```
---

class: inverse, center, middle

# I'm now able 
# to run algorithms
# on 100GB of data

---

## R Packages

```{r, out.width='100%'}
knitr::include_graphics("recap-packages.svg")
```

<br/>

### Paper in preparation: "Efficient management and analysis of large-scale genome-wide data with two R packages: bigstatsr and bigsnpr".

---

## Contributors are welcomed!

```{r, out.width='80%'}
knitr::include_graphics("https://applicationsandroidfrance.com/wp-content/uploads/2012/06/cat1200.jpg")
```

---

class: inverse, center, middle

# Thanks!

<br/><br/>

Package's website: https://privefl.github.io/bigstatsr/

Twitter and GitHub: [@privefl](https://twitter.com/privefl)

Presentation available online: https://goo.gl/nNg0hw

.footnote[Slides created via the R package [**xaringan**](https://github.com/yihui/xaringan).]
