<!DOCTYPE html>
<html>
  <head>
    <title>The R package bigstatsr: Memory- and Computation-Efficient Tools for Big Matrices</title>
    <meta charset="utf-8">
    <meta name="author" content="Florian Privé" />
    <link href="libs/remark-css/example.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# The R package <strong>bigstatsr</strong>:<br/>Memory- and Computation-Efficient Tools<br/>for Big Matrices
## useR!2017 lightning talk
### Florian Privé
### July 6, 2017

---




## About

I'm a PhD Student (2016-2019) in **Predictive Human Genetics** in Grenoble.

`$$\boxed{\Large{\text{Disease} \sim \text{DNA mutations}}}$$`

&lt;img src="http://www.blacksheep-van.com/wp-content/uploads/2016/03/Blacksheep-location-de-van-ame%CC%81nage-a%CC%80-Grenoble-1.jpg" style="display: block; margin: auto;" /&gt;

---

## Very large genotype matrices

- currently: 15K x 300K, [celiac disease](http://www.nature.com/ng/journal/v42/n4/abs/ng.543.html)

- soon: 500K x 800K, [UK Biobank](https://doi.org/10.1371/journal.pmed.1001779)
 
&lt;img src="https://media.giphy.com/media/3o7bueyxGydy48Lwgo/giphy.gif" width="65%" style="display: block; margin: auto;" /&gt;

---

## Problem I had

&lt;img src="memory-problem.svg" width="100%" style="display: block; margin: auto;" /&gt;

---

## Solution I found: R package **bigmemory**

&lt;img src="http://www.brokenthorn.com/Resources/images/virtual-memory%5B1%5D.png" style="display: block; margin: auto;" /&gt;

.footnote[Michael J. Kane, John Emerson, Stephen Weston (2013).]

---

## Dependencies on **bigmemory**

&lt;img src="bigmemory-depends.png" width="883" style="display: block; margin: auto;" /&gt;

---

## Sparse linear models: **biglasso**

&lt;img src="https://raw.githubusercontent.com/YaohuiZeng/biglasso/master/vignettes/2016-11-20_vary_p_pkgs.png" width="70%" style="display: block; margin: auto;" /&gt;

.footnote[Zeng, Y., and Breheny, P. (2017).]

---

## Partial Singular Value Decomposition

&lt;img src="svd.png" width="65%" style="display: block; margin: auto;" /&gt;

.footnote[based on R package **RSpectra**]

---

## Split-(par)Apply-Combine Strategy

### Apply standard R functions to big matrices (in parallel)

&lt;img src="split-apply-combine.svg" width="100%" style="display: block; margin: auto;" /&gt;

.footnote[strategy coined by Hadley Wickham (2011)]

---

## Test association of each column with an outcome

### In genetics, this is called a Genome-Wide Association Study (GWAS)

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="https://upload.wikimedia.org/wikipedia/commons/1/12/Manhattan_Plot.png" alt="Manhattan plot" width="100%" /&gt;
&lt;p class="caption"&gt;Manhattan plot&lt;/p&gt;
&lt;/div&gt;

---

## R Packages

&lt;img src="recap-packages.svg" width="100%" style="display: block; margin: auto;" /&gt;

---

class: inverse, center, middle

# I'm now able 
# to run algorithms
# on hundreds of 
# Gigabytes of data

---

## Any contributor is welcomed!

&lt;img src="https://applicationsandroidfrance.com/wp-content/uploads/2012/06/cat1200.jpg" width="80%" style="display: block; margin: auto;" /&gt;

---

class: inverse, center, middle

# Thanks!

&lt;br/&gt;&lt;br/&gt;

Package's website: https://privefl.github.io/bigstatsr/

Twitter and GitHub: [@privefl](https://twitter.com/privefl)

Presentation available online: https://goo.gl/nNg0hw

.footnote[Slides created via the R package [**xaringan**](https://github.com/yihui/xaringan).]
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": true
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {window.dispatchEvent(new Event('resize'));});</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
